inmemory-identity-provider/src/test/groovy/ru/ratauth/inmemory/ip/services/InMemorySessionServiceTest.groovy

```diff
@@ -0,0 +1,252 @@
+  void testGetByValidCodeShouldSuccess() {
+    Session session = createSession()
+    sessionService.create(session)
+
+    Session result = sessionService.getByValidCode(CODE, YESTERDAY).toBlocking().single()
+
+    assertEquals(result, session)
+  }
+
+  @Test
+  void testGetByValidCodeShouldThrowAuthCodeHasExpiredWhenSessionExpired() {
+    expectedException.expect(ExpiredException.class)
+    expectedException.expectMessage("Auth code has expired")
+    Session session = createSession()
+    sessionService.create(session)
+
+    sessionService.getByValidCode(CODE, TOMORROW).toBlocking().single()
+  }
+
+  @Test
+  void testGetByValidRefreshTokenShouldSuccess() {
+    Session session = createSession()
+    sessionService.create(session)
+
+    Session result = sessionService.getByValidRefreshToken(REFRESH_TOKEN, YESTERDAY)
+        .toBlocking()
+        .single()
+
+    assertEquals(result, session)
+  }
+
+  @Test
+  void testGetByValidRefreshTokenShouldThrowAuthCodeHasExpiredWhenSessionExpired() {
+    expectedException.expect(ExpiredException.class)
+    expectedException.expectMessage("Auth code has expired")
+    Session session = createSession()
+    sessionService.create(session)
+
+    sessionService.getByValidRefreshToken(REFRESH_TOKEN, TOMORROW).toBlocking().single()
+  }
+
+  @Test
+  void testGetByValidSessionTokenShouldSuccess() {
+    Session session = createSession()
+    sessionService.create(session)
+
+    Session result = sessionService.getByValidSessionToken(SESSION_TOKEN, YESTERDAY)
+        .toBlocking()
+        .single()
+
+    assertEquals(result, session)
+  }
+
+  @Test
+  void testGetByValidSessionTokenShouldThrowAuthCodeHasExpiredWhenSessionExpired() {
+    expectedException.expect(ExpiredException.class)
+    expectedException.expectMessage("Auth code has expired")
+    Session session = createSession()
+    sessionService.create(session)
+
+    sessionService.getByValidSessionToken(SESSION_TOKEN, TOMORROW).toBlocking().single()
+  }
+
+  @Test
+  void testGetByValidTokenShouldSuccess() {
+    Session session = createSession()
+    sessionService.create(session)
+
+    Session result = sessionService.getByValidToken(TOKEN, YESTERDAY)
+        .toBlocking()
+        .single()
+
+    assertEquals(result, session)
+  }
+
+  @Test
+  void testGetByValidTokenShouldThrowAuthCodeHasExpiredWhenSessionExpired() {
+    expectedException.expect(ExpiredException.class)
+    expectedException.expectMessage("Auth code has expired")
+    Session session = createSession()
+    sessionService.create(session)
+
+    sessionService.getByValidToken(TOKEN, TOMORROW).toBlocking().single()
+  }
+
+  @Test
+  void testAddEntrySuccess() {
+    Session session = createSession()
+    AuthEntry authEntry = new AuthEntry(authCode: "789",
+        relyingParty: CLIENT_NAME,
+        scopes: ['rs.read'] as Set,
+        refreshToken: REFRESH_TOKEN,
+    )
+
+    sessionService.create(session)
+    assertEquals(1, session.getEntries().size())
```

Comment:
Service must not modify input value in observable chain. All object that is passed through chain must be immutable or cloned with changed data. Seems like you implemented the wrong behaviour and test it here.